docs/RFCS/raft_consistency_checker.md

```diff
@@ -0,0 +1,86 @@
+- Feature Name: Raft consistency checker
+- Status: draft
+- Start Date: 2016-02-10
+- Authors: Ben Darnell, Bram Grunier, Vivek Menezes
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #837
+
+# Summary
+
+An online system that checks that periodically checks that range replicas
+are consistent; a snapshot for a range at a specific point in the raft log
+ has the same sha on all replicas.
+
+# Motivation
+
+Consistency! Correctness at scale.
+
+# Detailed design
+
+The proposal is for each range replica leader to periodically run the
+consistency checker. The consistency checker runs in three phases:
+
+1. All replicas agree on the snapshot at which a sha is to be computed.
+2. A sha is computed on all replicas.
+3. The leader shares its sha with all replicas; a replica computing a
+different sha from the one supplied by the leader alerts and eutanizes
+the node.
+
+The above is implemented by introducing two raft commands:
+
+1. CHECK_CONSISTENCY_BEGIN (BEGIN): A replica on receiving this command
+immediately acknowledges it has executed it, and asynchronously computes the sha
+for the current snapshot. It continues executing commands while it computes the
+sha on the side. Once the sha is computed it compares the BEGIN command id along
+with the computed sha to the one received from the leader in the second command
+below.
+2. CHECK_CONSISTENCY_CHECKSUM (CHECKSUM): This command contains the id for the
+BEGIN command it is referring to, and the sha that it has computed at the BEGIN
+command snapshot on the leader. On receiving this command the replica saves the
+command id and sha supplied, and compares the sha against the one it eventually
+computes. Under the rare circumstances that it sees a difference, it will alert
+and euthanize the node.
```

Comment:
sounds dangerous. If the leader is corrupted, every other node will kill itself? There needs to be some kind of quorum here.